README.md
# salah.rb

**Table of contents**

* [Introduction](#introduction)
* [Examples](#examples)
* [Install](#install)
* [Thanks](#thanks)
* [License](#license)

## <a id='introduction'>Introduction</a>

Salah.rb is a Ruby library that provides an API for retrieving Islamic prayer times.
The library is implemented by making HTTP requests to the [PrayerTimes.date](https://prayertimes.date)
web API. The [PrayerTimes.date](https://prayertimes.date) web API is free to use and at the time of
writing does not require an API key.

It's possible that in the future an API key will be required. If that comes to be you
can pass the `key: '...'` option to the methods described in this document.

## <a id='examples'>Examples</a>

**1.**

You can retrieve the prayer times for today, tomorrow, or this week using the methods
`Salah.today`, `Salah.tomorrow`, and `Salah.this_week`. You can provide either a city
name, the long/lat coordinates, or an IP address to determine your location.

The examples use `Salah.today` but you could also repeat them with `Salah.tomorrow` and
`Salah.this_week`.

**1.1** With a city name

```
> Salah.today(city: 'Amsterdam')
# => #<Salah::Response:0x000055e96db10560 @code="200", @status="OK", @prayers=[..]>
```

**1.2** With latitude and longitude coordinates

```
> Salah.today(latitude: 21.422510, longitude: 39.826168)
# => #<Salah::Response:0x000055e96db10560 @code="200", @status="OK", @prayers=[..]>
 ```

**1.3** With a Dutch IP address

```
> Salah.today(ip: '31.151.143.105')
# => #<Salah::Response:0x000055e96db10560 @code="200", @status="OK", @prayers=[..]>
```

**2.** `Salah.day`

The `Salah.day` method can request prayers for a given iso8601 date. The rest of the
method arguments are the same as `Salah.{today,tomorrow,this_week}`.

```
> Salah.day('2021-12-1', city: 'Amsterdam')
# => #<Salah::Response:0x...>
```

**3.** `Salah::Response`

**3.1** The response object

By now you might have noticed the `Salah::Response` object. It is returned by `Salah.today`,
`Salah.tomorrow`, `Salah.this_week` and `Salah.day`. Its most useful property is `Response#prayers`,
which returns an array of [`Salah::Prayer`](https://github.com/rg-3/salah.rb/blob/master/lib/salah/prayer.rb)
objects.

```ruby
res  = Salah.today(city: 'Cairo')
fajr = res.prayers.find(&:fajr?)
puts "Fajr is at #{fajr.time} in Cairo"
```

**3.2** Filtering by day of the week

The `Salah.this_week` method returns a response with an array of `Salah::Prayer`
objects for the week but what if you want to filter to find only the prayers
for Friday ? No worries, we can filter to include only Friday's prayers.

```ruby
prayers = Salah.this_week(city: 'Paris').prayers
prayers.select(&:friday?).each do |prayer|
  puts "#{prayer.name} starts at #{prayer.time} on Friday"
end
```

**4.**

Prayer times are calculated according to a school. The default school is the Muslim World League.
If you would like to use a different school, you can provide the `school: ` option. A list of schools
can be found with `Salah.schools`.

**4.1** `Salah.schools`

This method returns an array of `Salah::School` objects:

```
> Salah.schools
[#<struct Salah::School id=0, name="Ithna Ashari", fajr_angle=16, isha_angle=14>,
 #<struct Salah::School id=1, name="University of Islamic Sciences, Karachi", fajr_angle=18, isha_angle=18>,
 #<struct Salah::School id=2, name="Islamic Society of North America", fajr_angle=15, isha_angle=15>,
 #<struct Salah::School id=3, name="Muslim World League", fajr_angle=18, isha_angle=17>,
 #<struct Salah::School id=4, name="Umm Al-Qura University, Mecca", fajr_angle=18.5, isha_angle=90>,
 #<struct Salah::School id=5, name="Egyptian General Authority of Survey", fajr_angle=18, isha_angle=18>,
 #<struct Salah::School id=7, name="Institute of Geophysics, University of Tehran", fajr_angle=17.7, isha_angle=14>,
 #<struct Salah::School id=8, name="Morocco", fajr_angle=19, isha_angle=18>,
 #<struct Salah::School id=9, name="Department of Islamic Advancement, Malaysia (JAKIM)", fajr_angle=20, isha_angle=18>,
 #<struct Salah::School id=10, name="Majlis Ugama Islam Singapura", fajr_angle=20, isha_angle=18>,
 #<struct Salah::School id=11, name="Union des Organisations Islamiques de France", fajr_angle=12, isha_angle=12>,
 #<struct Salah::School id=12, name="Turkey", fajr_angle=12, isha_angle=12>]
```

**4.2** Choosing a particular school

The method of calculation can be set by setting the `school: ` option. The school option can be an integer
from 0 to 12 (excluding 6) or a `Salah::School` object. The default school is the Muslim World League.

```
> Salah.today(latitude: 21.422510, longitude: 39.826168, school: 7)
# => #<Salah::Response:0x000055e96db10560 @code="200", @status="OK", @prayers=[..]>
 ```

**5.** Short form coordinate options

It is possible to provide short form options for the longitude and latitude
params as `lng` and `lat` respectively.

```ruby
> Salah.today(lng: XX.XX, lat: XX.XX)
# => Salah::Response
```

## <a id='install'> Install </a>

Salah.rb is available [as a RubyGem](https://rubygems.org/gems/salah.rb).

    gem install salah.rb

## <a id='thanks'> Thanks </a>

Thanks to the developers and people at [PrayerTimes.date](https://www.prayertimes.date)
for providing the free web API service that powers salah.rb.

Thanks to The Sustainer for guiding me towards working on this project.

## <a id='license'> LICENSE </a>

MIT. See [/LICENSE.txt](/LICENSE.txt)


CHANGELOG.md
# CHANGELOG

## Next

* Update docs.

* Add `Salah::Response#headers`.

## v1.0.1

* Update the gems' description

## v1.0.0

* Add README.md documentation for `Salah.day`.

## v0.11.0

* Accept short form `lat` and `lng` for the latitude and longitude params.

## v0.10.0

* Fix `Salah.tomorrow` by implementing it on top of `Salah.day`.

* Add `Salah.day` for retrieving prayers for a given day.

## v0.9.0

* Add day of the week predicates to `Salah::Prayer` and `Salah::Date`.

## v0.8.1

* `git mv readme.md README.md`
* Documented parts of the API and improved the README.md documentation

## v0.8.0

* Default to the `Muslim World League` as the method for calculating prayer times.

* Cast location and juristic using `Kernel#Integer`.

* Add `Salah::{Location,Juristic}#to_int`.

* Allow an instance of `Salah::Juristic` to be passed as the `juristic: ` option.

* Add `Salah.juristics`.

* Add `Salah::Juristic`.

## v0.7.1

* Fix `openstruct-from_hash.rb` dependency.

## v0.7.0

* Add `Salah::HTTP::NoLocationError` and raise it when location requirements are missing.

* Add custom `user-agent` when making requests to the [PrayerTimes.date](https:/www.prayertimes.date) API.

* Add `Salah::Prayer#location`, which returns an instance of `Salah::Location`.

* Add `Salah::Location`.

## v0.5.0

* Add `Salah::Response#code`.

* Add `Salah::Response#status`.

## v0.4.0

* Improve `Salah::HTTP`.

* Add `Salah::HTTP::BadContentTypeError` and raise it when the response
  has an unexpected content type.

## v0.3.0

* Improve `Salah::Response`.

* Add `Salah::HTTP::BadResponseError` and raise it when the response is
  anything other than `Net::HTTPOK`.

## v0.2.0

* Add `Salah::Prayer#{fajr,dhuhr,asr,maghrib,isha}?` predicates.

## v0.1.3

* Update gem description.

## v0.1.2

* Update gem description.

## v0.1.1

* Update `Salah::HTTP#initialize` to accept an API key.

* Update `Salah::HTTP` to make use of the API key provided to `Salah::HTTP#initialize`.


lib/salah.rb
class Salah
  require "openstruct-from_hash"
  require_relative "salah/prayer"
  require_relative "salah/date"
  require_relative "salah/school"
  require_relative "salah/location"
  require_relative "salah/juristic"
  require_relative "salah/http"
  require_relative "salah/response"
  require_relative "salah/version"

  #
  # @return [Array<Salah::Juristic>]
  #   Returns an array of {Salah::Juristic} objects.
  #
  def self.juristics
    @juristics ||= [[0, "Shafii"], [1, "Hanafi"]].map { |params| Salah::Juristic.new(*params) }
  end

  #
  # @return [Array<Salah::School>]
  #   Returns an array of {Salah::School} objects.
  #
  def self.schools
    @schools ||= [
      [0, "Ithna Ashari", 16, 14],
      [1, "University of Islamic Sciences, Karachi", 18, 18],
      [2, "Islamic Society of North America", 15, 15],
      [3, "Muslim World League", 18, 17],
      [4, "Umm Al-Qura University, Mecca", 18.5, 90],
      [5, "Egyptian General Authority of Survey", 18, 18],
      [7, "Institute of Geophysics, University of Tehran", 17.7, 14],
      [8, "Morocco", 19, 18],
      [9, "Department of Islamic Advancement, Malaysia (JAKIM)", 20, 18],
      [10, "Majlis Ugama Islam Singapura", 20, 18],
      [11, "Union des Organisations Islamiques de France", 12, 12],
      [12, "Turkey", 12, 12]
    ].map { |params| Salah::School.new(*params) }
  end

  #
  # @return (see Salah::HTTP#today)
  #
  def self.today(options = {})
    new(key: options.delete(:key)).today(options)
  end

  #
  # @return (see Salah::HTTP#tomorrow)
  #
  def self.tomorrow(options = {})
    new(key: options.delete(:key)).tomorrow(options)
  end

  #
  # @return (see Salah::HTTP#this_week)
  #
  def self.this_week(options = {})
    new(key: options.delete(:key)).this_week(options)
  end

  #
  # @param (see Salah::HTTP#day)
  #
  # @return (see Salah::HTTP#day)
  #
  def self.day(date, options = {})
    new(key: options.delete(:key)).day(date, options)
  end

  def initialize(key: nil)
    @client = Salah::HTTP.new(key: key)
  end

  #
  # @return (see Salah::HTTP#today)
  #
  def today(options = {})
    @client.today(options)
  end

  #
  # @return (see Salah::HTTP#tomorrow)
  #
  def tomorrow(options = {})
    @client.tomorrow(options)
  end

  #
  # @return (see Salah::HTTP#this_week)
  #
  def this_week(options = {})
    @client.this_week(options)
  end

  #
  # @param (see Salah::HTTP#day)
  #
  # @return (see Salah::HTTP#day)
  #
  def day(date, options = {})
    @client.day(date, options)
  end
end
